### DISCUSSION TITLE
tflint --recursive --call-module-type=all cannot find nested local modules

### DISCUSSION THREAD

**Discussion by mdowling17** [terraform-linters/tflint#2082] [Q&A] [answered]
### Summary

When running `tflint --recursive --call-module-type=all` with a file directory that contains nested local modules (that is local modules that call other remote/local modules), tflint says `Error: "module1" module is not found. Did you run "terraform init"?` It seems to expect that I go into each of my local modules and run `terraform init`. That seems totally unnecessary since it should be able to figure things out by looking at my root .terraform/modules folder created by running terraform init in the root.

### Command

tflint --recursive --call-module-type=all

### Terraform Configuration

```terraform
$ tree
.
├── README.md
├── backend.tf
├── locals.tf
├── main.tf
├── modules
│   ├── module1
│   │   ├── main.tf
│   │   └── outputs.tf
│   ├── module2
│   │   ├── main.tf
│   │   └── variables.tf
│   ├── module3
│   │   ├── backend.tf
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   └── variables.tf
│   ├── module4
│   │   ├── main.tf
│   │   └── variables.tf
│   ├── module5
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   └── variables.tf
│   ├── module6
│   │   ├── main.tf
│   │   └── outputs.tf
│   └── module7
│       ├── main.tf
│       ├── outputs.tf
│       └── variables.tf
├── moved.tf
└── variables.tf
```

### TFLint Configuration

```HCL
plugin "terraform" {
  enabled = true
  preset  = "recommended"
}

config {
  call_module_type = "all"
}

rule "terraform_required_version" {
  enabled = false
}

rule "terraform_required_providers" {
  enabled = false
}
```

### Output

```Shell
│ Failed to run in modules/module7; exit status 1

Failed to load configurations; modules/module7/main.tf:23,1-33: "module8" module is not found. Did you run "terraform init"?; :

Error: "module8" module is not found. Did you run "terraform init"?

  on modules/module7/main.tf line 23, in module "module8":
  23: module "module8" {


│ Failed to run in modules/module3; exit status 1

Failed to load configurations; modules/module3/main.tf:14,1-16: "module9" module is not found. Did you run "terraform init"?; , and 1 other diagnostic(s):

Error: "module9" module is not found. Did you run "terraform init"?

  on modules/module3/main.tf line 14, in module "module9":
  14: module "module9" {

Error: "module10" module is not found. Did you run "terraform init"?

  on modules/module3/main.tf line 20, in module "module10":
  20: module "module10" {
```

### TFLint Version

0.51.2

### Terraform Version

1.5.7

### Operating System

- [ ] Linux
- [X] macOS
- [ ] Windows

**Comment by bendrucker** [comment]
If you recurse, those modules that you recurse into need to be properly initialized. Recursing and child module calls are entirely different things. Child modules have a formal tree relationship. Recursion assumes a tree structure of directories, but makes no logical assumptions about the relationships of those directories. Every recursed directory is an independent entity. When you've recursed into your child directories, there is no relationship that TFLint sees between your repository root and the module directory at that point.

**Reply by mdowling17** [reply]
But if I set it to `tflint --recursive --call-module-type=local` it works as expected without having to initialize each local module. The limitation with that approach, I presume, is that now it won't inspect remote modules. In my regular terraform workflow I would never run terraform init in each local module, rather I'd just run it in the root and it will generate all the submodules relative to the root and terraform plan works just fine. It'd be nice if tflint could do the same and simply refer to the root .terraform/modules folder to understand the tree structure.

**Reply by wata727** [reply]
For local modules there is a special optimization that allows child modules to be loaded without `.terraform/modules`. This is why `--call-module-type=local` works without `terraform init`.

Initially, I was planning on providing a solution to `--recursive` + `--call-module-type=all` issue, which would be to provide a `terraform init` hook that would be run in each subdirectory https://github.com/terraform-linters/tflint/issues/1930, but it seems interesting that maybe `terraform init` at the root module would be able to load the remote modules we need if we could resolve their addresses.

The issue here is that not all subdirectories are referenced from the root module. If the `module7` is not referenced from the root module, remote modules contained in it may not be included in the root `.terraform/modules`. In this case, the difficult part seems to be how to resolve the child modules when inspecting the `module7` by `--recursive`.

**Comment by h2oearth** [comment]
Hi @mdowling17 , did you end up finding how to fix this issue?

**Reply by mdowling17** [reply, marked as answer]
Yes, as I described above, if I set use `tflint --recursive --call-module-type=local` it works as expected. The limitation is that it doesn't lint remote modules, which I actually didn't need for my use-case. However, if you want to lint everything, you must go into each module directory and run terraform init and then you can run `tflint --recursive --call-module-type=all` from the root.

**Reply by h2oearth** [reply]
Thank you @mdowling17 , unfortunately, that does not work on my end :-/. I am using the latest tflint, terraform (1.95) and aws provider (5.66).
